rapid7_insight_agent/icon_rapid7_insight_agent/util/graphql_api/api_connection.py

```diff
@@ -0,0 +1,234 @@
+
+        results_object = result.json()
+
+        if results_object.get("errors"):
+            raise PluginException(cause="Insight Agent API returned errors",
+                                  assistance=results_object.get("errors"))
+
+        return results_object
+
+    def _setup_endpoint(self, region_string):
+        region_code = region_map.get(region_string)
+
+        if region_code:
+            endpoint = f"https://{region_code}.api.insight.rapid7.com/graphql"
+        else:
+            # It's an enum, hopefully this never happens.
+            raise PluginException(cause="Region not found.",
+                                  assistance="Region code was not found for selected region. Please contact support.")
+
+        return endpoint
+
+    def _get_all_agents(self):
+        """
+        Gets all available agents from the API
+
+        :return: list (agent objects)
+        """
+        agents = []
+        payload = {
+            "query": "query($orgId: String!) {organization(id: $orgId) { assets(first: 10000) { pageInfo { hasNextPage endCursor } edges { node { host { id vendor version description hostNames { name } primaryAddress { ip mac } uniqueIdentity { source id } attributes { key value } } id agent { id } } } } } } ",
+            "variables": {
+                "orgId": self.org_key
+            }
+        }
+
+        self.logger.info(f"Getting all agents...")
+        results_object = self._post_payload(payload)
+
+        has_next_page = results_object.get("data").get("organization").get("assets").get("pageInfo").get("hasNextPage")
+        agents.extend(self._get_agents_from_result_object(results_object))
+        self.logger.info(f"Initial agents received.")
+
+        # See if we have more pages of data, if so get next page and append until we reach the end
+        while has_next_page:
+            has_next_page, results_object = self._get_next_page_of_agents(agents, results_object)
+
+        self.logger.info(f"Done getting all agents.")
+
+        return agents
+
+    def _get_next_page_of_agents(self, agents, results_object):
+        """
+        In the case of multiple pages of returned agents, this will go through each page and append
+        those agents to the agents list
+
+        :param agents: list (agent objects)
+        :param results_object: dict
+        :return: tuple (boolean, dict (results object))
+        """
+        self.logger.info(f"Getting next page of agents.")
+        next_cursor = results_object.get("data").get("organization").get("assets").get("pageInfo").get("endCursor")
+        payload = {
+            "query": "query( $orgId:String! $nextCursor:String! ) { organization(id: $orgId) { assets( first: 1 after: $nextCursor ) { pageInfo { hasNextPage endCursor } edges { node { host { id vendor version description hostNames { name } primaryAddress { ip mac } uniqueIdentity { source id } attributes { key value } } id agent { id } } } } } }",
```

Comment:
```suggestion
            "query": "query( $orgId:String! $nextCursor:String! ) { organization(id: $orgId) { assets( first: 1 after: $nextCursor ) { pageInfo { hasNextPage endCursor } edges { node { host { id vendor version description hostNames { name } primaryAddress { ip mac } uniqueIdentity { source id } attributes { key value } } id agent { id } } } } } }",
```
I see that this query has "first: 1" instead of "first: 10000" like the initial one - is this a typo?

Reply:
Also it looks like the query is very similar to the initial query - could you use f strings and recursion to turn this into a single function?

Reply:
I tried that (f strings). It gets really complicated because these have '{' in them. So if I use an f string, I end up having to escape all the brackets, and that gets unreadable very quickly.

Reply:
I fixed the 1 vs 10000 typo. 

The problem with f strings here are all the '{' ... Escaping this to an f string will look awful. You have to double { ...so 

some_python_var = "value"
"{{ {some_python_var} }}" If you want "{ value }"

I tried a couple ways of if injecting "quarantine" and "unquarantine" but they just didn't pan out.